### ISSUE TITLE
Analysis: React Accessibility Rules

### ISSUE THREAD

**Issue by HamletDRC** [Microsoft/tslint-microsoft-contrib#188]
Microsoft is in the process of creating some React based accessibility tslint rules. We had a small internal brainstorm session. @ipip2005 mentioned in a related issue that he/her was interested as well. 

Our current plans are: 

Images must have alt-text 
* reference - https://www.w3.org/TR/WCAG10/wai-pageauth.html#tech-text-equivalent
* including image buttons - http://webaim.org/techniques/forms/controls#image_button
* including images used as submit buttons - <INPUT type="image" name=submit src="button.gif" alt="Submit">

Labels must be explicitly associated with their controls
* reference - https://www.w3.org/TR/WAI-WEBCONTENT-TECHS/#tech-associate-labels
* more acceptable examples - http://www.clarissapeterson.com/2012/11/html5-accessibility/
* Use title when label not available: https://www.w3.org/TR/2015/NOTE-WCAG20-TECHS-20150226/H65

Describe the purpose of frames
* reference: https://www.w3.org/TR/WCAG10/wai-pageauth.html#tech-frame-longdesc

Form buttons must have a descriptive value
* reference - http://webaim.org/standards/wcag/checklist
* for examples see http://webaim.org/techniques/forms/controls#button

Form inputs must have associated text labels 
* reference - http://webaim.org/standards/wcag/checklist

Text inputs and TextArea inputs must be labeled
* reference - http://webaim.org/techniques/forms/controls#input

Elements with onClick listeners must define role="button" 
* reference - https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_button_role

Required input elements must define aria-required="true" 
* reference - http://www.clarissapeterson.com/2012/11/html5-accessibility/

Edit boxes and text areas must include default, place-holding characters
* https://www.w3.org/TR/WCAG10/wai-pageauth.html#tech-place-holders
* https://www.w3.org/TR/WCAG10-HTML-TECHS/#forms-specific

Images used as "submit" buttons must have text equivalent
* https://www.w3.org/TR/WCAG10-HTML-TECHS/#forms-specific

Tables must have summaries
* tables should have either a caption or a title: https://www.w3.org/TR/WCAG10-HTML-TECHS/#table-summary-info
* tables should have a summary attribute: https://www.w3.org/TR/WCAG10-HTML-TECHS/#table-summary-info
* Provide terse substitutes for header labels with the "abbr" attribute on TH

Do not use obsolete html elements/attributes
* https://www.w3.org/TR/html5/obsolete.html

Is there anything else you'd like to see? 
Any comments? 

Thanks in advance!

**Comment by HamletDRC** [comment]
THere is an es6 related repo here: https://github.com/evcohen/eslint-plugin-jsx-a11y

**Comment by ipip2005** [comment]
I'm also writing rules based on this repo. Personally I would prefer to see these rules at first
aria-role, aria-props, aria-proptypes, tabindex-no-positive, role-has-required-aria-props, role-supports-aria-props, img-has-alt

These rules checks a11y fundamentals.

And for me, some rules from https://github.com/evcohen/eslint-plugin-jsx-a11y are not so easy to be used by projects because they will easily break their projects and people might not very likely to fix them. For example
Required input elements must define aria-required="true"
Elements with onClick listeners must define role="button"

**Comment by HamletDRC** [comment]
I sent an email. It would be great to have these rules in the tslint-microsoft-contrib repo.

**Comment by HamletDRC** [comment]
Closing this issue because many child issues have been created
